uplink/piecestore/upload.go

```diff
@@ -138,7 +142,13 @@ func (client *Upload) Write(data []byte) (written int, err error) {
 			},
 		})
 		if err != nil {
-			err = ErrProtocol.Wrap(err)
+			_, closeErr := client.stream.CloseAndRecv()
+			if err != io.EOF {
+				err = ErrProtocol.Wrap(errs.Combine(err, closeErr))
+			} else {
+				err = ErrProtocol.Wrap(closeErr)
```

Comment:
If `err == io.EOF` and `closeErr == nil` then this wouldn't return `nil` as the error.